sunscreen_tfhe/src/math/simd/avx2.rs

```diff
@@ -0,0 +1,150 @@
+use num::Complex;
+use raw_cpuid::CpuId;
+
+#[cfg(target_arch = "x86")]
+use core::arch::x86::*;
+
+#[cfg(target_arch = "x86_64")]
+use core::arch::x86_64::*;
+use std::{arch::asm, sync::OnceLock};
+
+use super::scalar;
+
+#[inline(always)]
+fn avx_512_available() -> bool {
+    static AVX512_AVAILABLE: OnceLock<bool> = OnceLock::new();
+
+    *AVX512_AVAILABLE.get_or_init(|| {
+        if let Some(e) = CpuId::new().get_extended_feature_info() {
+            e.has_avx512f()
+        } else {
+            false
+        }
+    })
+}
+
+#[inline(always)]
+/// Compute vector `c += a * b` over &[Complex<f64>].
+///
+/// # Panics
+/// If `c.len() != a.len() != b.len()`
+/// If `a.len() % 8 != 0`
+/// If `a`, `b`, `c` are not aligned to a 512-bit boundary.
+pub fn complex_mad(c: &mut [Complex<f64>], a: &[Complex<f64>], b: &[Complex<f64>]) {
+    assert_eq!(c.as_ptr().align_offset(core::mem::align_of::<__m512d>()), 0);
+    assert_eq!(b.as_ptr().align_offset(core::mem::align_of::<__m512d>()), 0);
+    assert_eq!(a.as_ptr().align_offset(core::mem::align_of::<__m512d>()), 0);
+    assert_eq!(c.len(), a.len());
+    assert_eq!(b.len(), a.len());
+    assert_eq!(a.len() % 8, 0);
+
+    if avx_512_available() {
+        unsafe { complex_mad_avx_512_unchecked(c, a, b) }
+    } else {
+        scalar::complex_mad(c, a, b)
+    }
+}
+
+/// Compute vector `c += a * b` over `&[Complex<f64>]`.
+/// This function is very unsafe.
+///
+/// # Safety
+/// c, a, b must be aligned to a 512-bit boundary. The program will otherwise bus error
+/// and crash.
+///
+/// the lengths of c, a, and b must be the equal or UB may result.
+/// the lengths of c, a, and b must be a multiple of 8 or UB will result.
+#[inline(always)]
+unsafe fn complex_mad_avx_512_unchecked(
```

Comment:
Potentially add a gated test function to ensure this matches on avx512f enabled machines.